### PR TITLE
✨Enhanced nodeports to support symlinks

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

The sidecar will mount inside ´/dy-volumes´ the full path of the volume. Examples:
- service requests to a volume for `/tmp/data/state` -> dynamic-sidecar will mount to `/dy-volumes/tmp/data/state`
- service requests to a volume for `/www/inputs` -> dynamic-sidecar will mount to `/dy-volumes/www/inputs`
- service requests to a volume for `/www/outputs` -> dynamic-sidecar will mount to `/dy-volumes/www/outputs`

Basically the dynamic-sidecar treats ´/dy-volumes´ as the root folder ´/´ of the service. Now, if a user will try to create a symlink, it will now have all the required relative paths.

- Allows nodeports to use symlinks.
- Out of the box, nodeports will upload the content pointed by the symlink.

## Checklist

- [x] Unit tests for the changes exist
- [x] Runs in the swarm

### CODE REVIEW

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py [hunk 1]
```diff
@@ -159,9 +160,25 @@ async def _set(self, new_value: ItemConcreteValue) -> None:
...
 
-            if isinstance(converted_value, Path):
-                if not converted_value.exists() or not converted_value.is_file():
+            def _raise_if_file_not_on_disk(path: Path) -> None:
+                if not path.exists() or not path.is_file():
+                    log.warning("Could not find a valid file %s", path)
                     raise InvalidItemTypeError(self.property_type, str(new_value))
+
+            if isinstance(converted_value, Path):
+                if converted_value.is_symlink():
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:169]: IMO following links should be optional (SEE  discussion about ``follow_symlinks`` in [os](https://docs.python.org/3/library/os.html) ) 

A situation like ``workdir``

```console
$ ls -la workdir
drwxrwxr-x 2 scu scu 4096 Nov 10 22:21 .
drwxrwxr-x 8 scu scu 4096 Nov 10 22:21 ..
-rw-rw-r-- 1 scu scu    0 Nov 10 22:21 file.dat
lrwxrwxrwx 1 scu scu   11 Nov 10 22:21 inputs -> /data/inputs
lrwxrwxrwx 1 crespo crespo   12 Nov 10 22:21 outputs -> /data/outputs
```
I would like to upload to s3 ``workdir`` but avoid ``inputs`` and ``outputs``. But when downloaded, I would like the symlinks to be restored

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py [hunk 2]
```diff
@@ -159,9 +160,25 @@ async def _set(self, new_value: ItemConcreteValue) -> None:
...
+                    symlink_target_path = Path(os.readlink(converted_value))
+                    if not symlink_target_path.is_absolute():
+                        # in the case of a relative symlink
+                        # since the working directory is different form the
+                        # one in which the symlink is placed,
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:174]: how do you account for recursive links? i.e. a link to a folder within the ``workdir`` that you are uploading

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/validation.py [hunk 3]
```diff
@@ -203,6 +203,11 @@ async def validate_compose_spec(
             settings_env_vars=service_settings_env_vars,
         )
 
+        # FIXME: tmp to comply with
```
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/validation.py:206]: this has nothing to do with this PR :-)

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py (hunk 2 above)
Comment (GitHK) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:174]: one symlinks to files are supported, function `_raise_if_file_not_on_disk` ensures the symlink points to a file on the disk.

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py (hunk 1 above)
Comment (pcrespov) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:169]: - in nodeports config
- nodeports will ultimately be used almost exclusively in the dy-sidecar
- it is up the the service to decide ``follow_symlinks``
- I would rather default ``follow_symlinks=False`` as in the os lib

Comment (GitHK) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:169]: If you look carefully, `follow_symlinks=False` is Falso only when dealing with permission changes and some other border line cases. When the file needs to be read/written `follow_symlinks=True`.
That is exactly our case. I would default it to `True`
I agree into storing the configuration into nodeports.

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py [hunk 4]
```diff
@@ -77,18 +78,51 @@ def e_tag() -> str:
...
-def this_node_file(tmp_path: Path) -> Path:
-    file_path = this_node_file_name()
-    file_path.write_text("some dummy data")
+@pytest.fixture(
+    params=[
+        True,
+        False,
+    ]
+)
+def is_symlink(request) -> bool:
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:87]: this fixture name is pretty unclear. are we allowing for symlink? or are we creating a symlink?

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py [hunk 5]
```diff
@@ -77,18 +78,51 @@ def e_tag() -> str:
...
+
+    file_name = this_node_file_name()
+
+    symlink_path: Optional[Path] = None
+    return_path: Optional[Path] = None
+
+    if is_symlink:
+        symlink_path = file_name
+        file_path = file_name.parent / f"source_{file_name.name}"
+        file_path.write_text("some dummy data")
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:103]: you might want to separate that fixture in 2, you might even just use the tempfile fixture for that...
in case it's a symlink you return the symlink, and if not you return the file itself.

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py [hunk 6]
```diff
@@ -546,6 +580,7 @@ class FakeNodePorts:
         user_id: int
         project_id: str
         node_uuid: str
+        follow_symlinks: bool = True
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:583]: why True?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py [hunk 7]
```diff
@@ -91,12 +90,12 @@ async def get_dynamic_sidecar_spec(
     #
     # Two separate volumes are required to achieve the following on the spawned
     # dynamic-sidecar containers:
-    #   `volume_name_inputs:/target_folder/inputs`
-    #   `volume_name_outputs:/target_folder/outputs`
-    #   `volume_name_path_to_sate_01:/target_folder/path_to_sate_01`
+    #   `volume_name_path_to_inputs:/target_folder/path/to/inputs`
+    #   `volume_name_path_to_outputs:/target_folder/path/to/outputs`
+    #   `volume_name_path_to_sate_01:/target_folder/path/to/sate/01`
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py:95]:
```suggestion
    #   `volume_name_path_to_sate_01:/target_folder/path/to/state/01`
```

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py [hunk 8]
```diff
@@ -52,21 +52,27 @@ def test_expected_paths(
     state_paths: List[Path],
     expect: Callable[[str, str], Dict[str, Any]],
 ) -> None:
-    path = Path("/inputs")
+    inputs_path = Path("/tmp/some/inputs")
```
Comment (sanderegg) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:55]: please use the tmp_folder fixture for this

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py [hunk 9]
```diff
@@ -52,21 +52,27 @@ def test_expected_paths(
...
-        compose_namespace, path, node_uuid
-    ) == expect(source=f"{compose_namespace}_inputs", target="/dy-volumes/inputs")
+        compose_namespace, inputs_path, node_uuid
+    ) == expect(
+        source=f"{compose_namespace}_tmp_some_inputs",
+        target=str(Path("/dy-volumes") / f"{inputs_path}".strip("/")),
+    )
 
-    path = Path("/outputs")
+    outputs_path = Path("/tmp/some/outputs")
```
Comment (sanderegg) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:63]: same here

File: services/dynamic-sidecar/tests/unit/test_modules_mounted_fs.py [hunk 10]
```diff
@@ -69,16 +71,28 @@ def test_expected_paths_and_volumes(
...
+        == mounted_fs.DY_VOLUMES / f"{outputs_dir}".strip("/")
+    )
+    assert (
+        mounted_volumes.disk_inputs_path
+        == mounted_fs.DY_VOLUMES / f"{inputs_dir}".strip("/")
+    )
 
     assert set(mounted_volumes.disk_state_paths()) == {
-        mounted_fs.DY_VOLUMES / _replace_slashes(x).strip("_") for x in state_paths_dirs
+        mounted_fs.DY_VOLUMES / f"{x}".strip("/") for x in state_paths_dirs
```
Comment (sanderegg) [inline services/dynamic-sidecar/tests/unit/test_modules_mounted_fs.py:84]: there must be some kind of clean function in pathlib no?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/volumes_resolver.py [hunk 11]
```diff
@@ -4,19 +4,17 @@
...
-
-    @classmethod
-    def _name_from_path(cls, path: Path) -> str:
-        return str(path).replace(os.sep, "_")
+    BASE_PATH: Path = Path("/dy-volumes")
 
     @classmethod
     def _target(cls, state_path: Path) -> str:
-        return f"{cls.BASE_PATH}/{cls._name_from_path(state_path).strip('_')}"
+        target_path = cls.BASE_PATH / f"{state_path}".strip("/")
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/volumes_resolver.py:11]:
```python
>>> Path("/ciasdja/asdasjasd/ads").relative_to("/")
PosixPath('ciasdja/asdasjasd/ads')
```
Does this look less like the '90?

[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/volumes_resolver.py:11]: @sanderegg

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 8 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:55]: no need for the folder here, only the name

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 9 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:63]: same as above

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 6 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:583]: By default we allow to follow symlinks. To disable this, it must put it `False`.  If a symlink is used in this and symlinks are disabled an error will be raised.

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 4 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:87]: `return_symlink` seems to be a more adequate name

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 5 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py]: Think I managed to make it more readable.

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/nodeports_v2.py [hunk 12]
```diff
@@ -27,6 +27,13 @@ class Nodeports(BaseModel):
         [DBManager, int, str, str], Coroutine[Any, Any, Type["Nodeports"]]
     ]
     auto_update: bool = False
+    follow_symlinks: bool = Field(
+        True,
+        description=(
+            "When uploading a file, in most cases the content should "
+            "be uploaded and. The symlink per se should not be stored"
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_v2/nodeports_v2.py:34]: TYPE: ``The symlink PAIR ...``

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/nodeports.py [hunk 13]
```diff
@@ -52,7 +52,7 @@ async def upload_outputs(outputs_path: Path, port_keys: List[str]) -> None:
     )
 
     # let's gather the tasks
-    temp_files: List[Path] = []
+    temp_paths: Deque[Path] = deque()
```
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/nodeports.py:55]: :-)

**Review by pcrespov — APPROVED** [review]

File: packages/simcore-sdk/src/simcore_sdk/node_ports_common/exceptions.py [hunk 14]
```diff
@@ -51,6 +51,18 @@ def __init__(self, item_type: str, item_value: str, msg: Optional[str] = None):
         self.item_value = item_value
 
 
+class SymlinksUsageIsDisabledError(NodeportsException):
+    """Item type incorrect"""
+
+    def __init__(self, msg: Optional[str] = None):
```
Comment (sanderegg) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_common/exceptions.py:57]: what is the point of having a msg here since you do not use it?

File: packages/simcore-sdk/tests/integration/test_node_ports_v2_nodeports2.py [hunk 15]
```diff
@@ -21,6 +22,10 @@
 from simcore_sdk.node_ports_v2.links import ItemConcreteValue
 from simcore_sdk.node_ports_v2.nodeports_v2 import Nodeports
 
+SYMLINK_PATH = Path(tempfile.gettempdir()) / f"symlink_{Path(__file__).name}"
+if not SYMLINK_PATH.exists():
+    os.symlink(__file__, SYMLINK_PATH)
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/integration/test_node_ports_v2_nodeports2.py:27]: I might be missing something here, but are you creating a symlink each time that module is read?
-> fixture with cleanup code ???

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py [hunk 16]
```diff
@@ -74,21 +77,66 @@ def e_tag() -> str:
     return "1212132546546321-1"
 
 
-##################### FIXTURES
+@contextmanager
+def symlink_to_file_with_data(file_name: Path) -> Iterator[Path]:
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:81]: I do not understand why you use a contextmanager here instead of a fixture? what does that bring?

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py [hunk 17]
```diff
@@ -74,21 +77,66 @@ def e_tag() -> str:
...
-def this_node_file(tmp_path: Path) -> Path:
-    file_path = this_node_file_name()
+    symlink_path.unlink()
+    assert not symlink_path.exists()
+    file_path.unlink()
+    assert not file_path.exists()
+
+
+@contextmanager
+def file_with_data(file_name: Path) -> Iterator[Path]:
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:101]: same here

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 6 above)
Comment (sanderegg) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:597]: The node ports is created dynamically. I would expect to setup such a setting (or its default there)
I do not understand why the test fixture has a default value. It kinds of defeat the purpose no?

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 8 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:55]: yes but you are still mounting things from a folder that might not always exist in every distro right? tmp_folder fixture would prevent such problems

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:63]: same as above

**Review by sanderegg — CHANGES_REQUESTED** [review]
a few small things to check.

File: packages/simcore-sdk/tests/integration/test_node_ports_v2_nodeports2.py (hunk 15 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/integration/test_node_ports_v2_nodeports2.py:27]: migrated to a fixture

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 16 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:81]: I managed to refactor it using fixtures

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 17 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:101]: see above

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 6 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:597]: You can set it up. When you instantiate nodeports, the default value is `True` in that case as well.

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 8 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:55]: You are correct. The director-v2 will never mount to that path. It is just used to send the information downstream to the dynamic-sidecar when it is started.
Also in the scope of this unit test nothing will get mounted as well.

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 9 above)
Comment (GitHK) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:63]: see above

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 6 above)
Comment (sanderegg) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py:586]: yes but why having a default in a test fixture? in a test you have to define things completely. we do not want things that are defined by default there, they generate confusion when we do not understand how it gets set later on.
I would definitely vote for removing any default in a test argument or else fix the other test cases that use that fixture.

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 8 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:55]: then a tip: you should use the faker if the path is totally irrelevant. it would make the intent clearer

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_volumes_resolver.py:63]: same as above

**Review by sanderegg — CHANGES_REQUESTED** [review]
Sorry to be a pain, but last time we had default values in tests/fixtures it took a while to debug. I would prefer tests without defaults of any sort.

File: packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py (hunk 6 above)
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_v2_port.py]: Removing this as agreed, also informed @pcrespov about it.

**Review by sanderegg — APPROVED** [review]

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py [hunk 18]
```diff
@@ -167,8 +167,6 @@ def _raise_if_file_not_on_disk(path: Path) -> None:
 
             if isinstance(converted_value, Path):
                 if converted_value.is_symlink():
-                    if not self._node_ports.follow_symlinks:
-                        raise SymlinksUsageIsDisabledError()
                     symlink_target_path = Path(os.readlink(converted_value))
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:170]: I think there was a misunderstanding here. My point was not that if follow_symlinks was true, to raise an exception ... there is no point in adding a variable if raises an exception if used. 

Aside of the name of the variable, the problem is that I see three feasible situations: 
1) stores the symlink reference and then restores the symlink back. (probably if the refereed data does not exist upon restoring then it should error)
2) stores the symlink referee (actual data) and restores the data. In this case, it ensures the data will be in place but will not be in sync anymore with the original.
3) some hybrid that stores both 1 and 2 and restores the data (if the original does not exists) or the symlink (if the original exists). 

I think you are implementing 2 but i can imagine some cases with 1.

Comment (GitHK) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:170]: I think this is a bit of an overstretch. Our use case is a simple one.

When taking about this feature, I think of nodeports, as follows: node A can upload to its output a file which can be used by whatever type of node that accepts a file as inputs. With this in mind:
1. If you want to support this then you should only store a symlink which points to file that also gets uploaded. If this is the case I would encourage the user to use directly the file name or keep this symlink somewhere else. **I would not support this case**.
2. When node A uploads a symlink it uploads the content of that symlink. Node B will download the contents as if it would be the original file. Node B does not care about the origin of the data, it knows that what it has is the latest version. **Yes this is the current use case**.
3. I did not understand it and perhaps is already too complicated.

Basically yes, I am implementing 2 but I do not want to support 1 because I see no added value since we can instruct the users to do something else.

Comment (pcrespov) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py:170]: ok. agreed to implement only 2. I have a use case for 1 and we will follow up on it. thx
